crates/encoding/src/path.rs

```diff
@@ -723,31 +763,25 @@ impl fello::scale::Pen for PathEncoder<'_> {
     }
 }
 
+const EPSILON: f32 = 1e-12;
+
 // Returns the end point of the start tangent of a curve starting at `(x0, y0)`, or `None` if the
 // curve is degenerate / has zero-length. The inputs are a sequence of control points that can
 // represent a line, a quadratic Bezier, or a cubic Bezier. Lines and quadratic Beziers can be
 // passed to this function by simply setting the invalid control point degrees equal to `(x0, y0)`.
 fn start_tangent_for_curve(
     p0: (f32, f32),
     p1: (f32, f32),
-    p2: (f32, f32),
-    p3: (f32, f32),
+    p2: Option<(f32, f32)>,
+    p3: Option<(f32, f32)>,
 ) -> Option<(f32, f32)> {
```

Comment:
Small style suggestion: making this a method of `PathEncoder` and dropping the `p0` parameter (since it always appears to be `PathEncoder::first_point`?) should simplify the call sites a bit.